spec/index.adoc

```diff
@@ -975,6 +975,28 @@ During these steps, no other entry point of the old or new canister is invoked.
 
 These steps are atomic: If `canister_pre_upgrade` or `canister_post_upgrade` trap, the upgrade has failed, and the canister is reverted to the previous state. Otherwise, the upgrade has succeeded, and the old instance is discarded.
 
+[#system-api-canister-eviction]
+==== Canister eviction
+
+Eviction is a special type of canister upgrade that skips execution of the `canister_pre_upgrade` method on the old canister instance.
+The main purpose of this mode is recovery from cases when the `canister_pre_upgrade` hook traps unconditionally preventing the normal upgrade path.
+
+NOTE: Canister eviction can lead to data loss. Use it only as the last resort and only if the stable memory already contains the entire canister state.
+
+The IC executes canister eviction as follows:
+
+ 1. Discard the old canister version except of its <<system-api-stable-memory,stable memory>>.
```

Comment:
canister version is now the integer bumped upon changes to the canister (e.g., upgrade or change of controllers); I'd refer here to "canister module and state" instead of "canister version"